### PR TITLE
Cache number of violations per build

### PR DESCRIPTION
**Pull request by gylaz**
We display the number of violations per repo on the account page (for
paid repos). That was causing timeouts, since repos can have many builds
and sql `IN` uses a loop for large arrays.
To speed things up, we can cache the number of violations (messages) per
build.

### CODE REVIEW

File: db/migrate/20151216235118_add_violations_count_to_builds.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddViolationsCountToBuilds < ActiveRecord::Migration
+  def change
+    add_column :builds, :violations_count, :integer
```
Comment (teoljungberg) [inline db/migrate/20151216235118_add_violations_count_to_builds.rb]: Should this default to `0`? Or does it to that by default?

File: spec/models/violation_spec.rb [hunk 2]
```diff
@@ -34,4 +34,15 @@
       expect(violation.messages_count).to eq 2
     end
   end
+
+  describe "#after_create" do
```
Comment (teoljungberg) [inline spec/models/violation_spec.rb]: This isn't necessarily a method we're testing, but a flow through the lifecycle of the object.

Would `context "after save callbacks"` be a better fit?

File: db/schema.rb [hunk 3]
```diff
@@ -56,10 +57,11 @@
...
   create_table "memberships", force: :cascade do |t|
-    t.integer  "user_id",    null: false
-    t.integer  "repo_id",    null: false
-    t.datetime "created_at", null: false
-    t.datetime "updated_at", null: false
+    t.integer  "user_id",                    null: false
+    t.integer  "repo_id",                    null: false
+    t.datetime "created_at",                 null: false
+    t.datetime "updated_at",                 null: false
+    t.boolean  "admin",      default: false
```
Comment (thorncp) [inline db/schema.rb]: What is this for? Seems unrelated to this PR.

File: db/migrate/20151216235118_add_violations_count_to_builds.rb (hunk 1 above)
Comment (gylaz) [inline db/migrate/20151216235118_add_violations_count_to_builds.rb]: Will adding a default here set all the existing columns to `0`? I wasn't sure and I wanted to avoid that, so that we know if it's `nil` it means it hasn't been calculated.

File: spec/models/violation_spec.rb (hunk 2 above)
Comment (gylaz) [inline spec/models/violation_spec.rb]: I like that idea, but maybe `describe` is a better fit, since `context` is better suited for documenting the special situation/condition.

File: db/schema.rb (hunk 3 above)
Comment (gylaz) [inline db/schema.rb]: Nice catch! This snuck in from a different branch/feature.

File: db/migrate/20151216235118_add_violations_count_to_builds.rb (hunk 1 above)
Comment (teoljungberg) [inline db/migrate/20151216235118_add_violations_count_to_builds.rb]: Fair reasoning, I can stand behind that

Comment (gylaz) [inline db/migrate/20151216235118_add_violations_count_to_builds.rb]: When do you sleep!?

**Comment by thorncp** [comment]
:+1:

File: db/migrate/20151216235118_add_violations_count_to_builds.rb (hunk 1 above)
Comment (teoljungberg) [inline db/migrate/20151216235118_add_violations_count_to_builds.rb]: Timezones are finicky.
